ldm/invoke/concepts_lib.py

```diff
@@ -0,0 +1,143 @@
+        self.concepts_loaded = dict()
+        self.triggers = dict()            # concept name to trigger phrase
+        self.concept_names = dict()       # trigger phrase to concept name
+
+    def list_concepts(self)->list:
+        '''
+        Return a list of all the concepts by name, without the 'sd-concepts-library' part.
+        '''
+        if self.concept_list is not None:
+            return self.concept_list
+        try:
+            models = self.hf_api.list_models(filter=ModelFilter(model_name='sd-concepts-library/'))
+            self.concept_list = [a.id.split('/')[1] for a in models]
+        except Exception as e:
+            print(' ** WARNING: Hugging Face textual inversion concepts libraries could not be loaded. The error was {str(e)}.')
+            print(' ** You may load .bin and .pt file(s) manually using the --embedding_manager argument.')
+        return self.concept_list
+
+    def get_concept_model_path(self, concept_name:str)->str:
+        '''
+        Returns the path to the 'learned_embeds.bin' file in
+        the named concept. Returns None if invalid or cannot
+        be downloaded.
+        '''
+        return self.get_concept_file(concept_name.lower(),'learned_embeds.bin')
+
+    def concept_to_trigger(self, concept_name:str)->str:
+        '''
+        Given a concept name returns its trigger by looking in the
+        "token_identifier.txt" file.
+        '''
+        if concept_name in self.triggers:
+            return self.triggers[concept_name]
+        file = self.get_concept_file(concept_name, 'token_identifier.txt')
+        if not file:
+            return None
+        with open(file,'r') as f:
+            trigger = f.readline()
+            trigger = trigger.strip()
+        self.triggers[concept_name] = trigger
+        self.concept_names[trigger] = concept_name
+        return trigger
+
+    def trigger_to_concept(self, trigger:str)->str:
+        '''
+        Given a trigger phrase, maps it to the concept library name.
+        Only works if concept_to_trigger() has previously been called
+        on this library. There needs to be a persistent database for
+        this.
+        '''
+        concept = self.concept_names.get(trigger,None)
+        return f'<{concept}>' if concept else f'{trigger}'
+
+    def replace_triggers_with_concepts(self, prompt:str)->str:
+        '''
+        Given a prompt string that contains <trigger> tags, replace these
+        tags with the concept name. The reason for this is so that the
+        concept names get stored in the prompt metadata. There is no
+        controlling of colliding triggers in the SD library, so it is
+        better to store the concept name (unique) than the concept trigger
+        (not necessarily unique!)
+        '''
+        def do_replace(match)->str:
+            return self.trigger_to_concept(match.group(1)) or f'<{match.group(1)}>'
+        return re.sub('(<[^>]+>)', do_replace, prompt)
```

Comment:
duplication of this regex in four separate places over two python source fails breaks both the DRY (Don't Repeat Yourself) principle in two different ways at the same time

Reply:
Fixed. I only count the regexp being repeated twice, and in fact the two instances are slightly different (one includes the angle brackets in the match group and the other doesn't). Where are the other two instances?